### PR TITLE
#15 transform

### LINKED ISSUE #15: Transform
# Description

This ticket creates a transform file that transforms the extracted data in preparation for loading to the RDS database.

# Outcome

We are able to transform the data into a relevant data through removing irrelevant data and processing the data for more insights such as sentiment.

# User story

As a data engineer, I want to transform the extracted data by cleaning and processing it (e.g., removing irrelevant data and deriving insights like sentiment), so that it is optimized and ready for loading into the RDS database.

### PR DESCRIPTION
**Pull request by keoghrmj**
Transform.py script works for a csv file and it creates a dataframe. This is currently finished for now

Closes #15

### CODE REVIEW

File: pipeline/transform.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    )
+    cursor = conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
+    return cursor, conn
+
+
+def clean_data(bluesky_data: pd.DataFrame, keywords: list[str]) -> pd.DataFrame:
+    """Removes any rows that don't include the keywords and removes duplicate rows."""
+
+    # filtered_bluesky = bluesky_data[bluesky_data['text'].str.contains(
+    #    r'\b(?:' + '|'.join(keywords) + r')\b', case=False, na=False, regex=True)]
```
Comment (SurinaCS) [inline pipeline/transform.py]: remove commented out code please!

File: pipeline/transform.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        database=config["DB_NAME"]
+    )
+    cursor = conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
+    return cursor, conn
+
+
+def clean_data(bluesky_data: pd.DataFrame, keywords: list[str]) -> pd.DataFrame:
+    """Removes any rows that don't include the keywords and removes duplicate rows."""
+
+    # filtered_bluesky = bluesky_data[bluesky_data['text'].str.contains(
```
Comment (Kurt812) [inline pipeline/transform.py:42]: Should these lines be commented out?

File: pipeline/transform.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+import logging
+import os
+import pandas as pd
+import psycopg2
+import psycopg2.extras
+from dotenv import dotenv_values
+from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
+
+
+# Load csv created by extract into relevant data structure e.g. pd.Dataframe
```
Comment (Kurt812) [inline pipeline/transform.py:14]: Some more comments to remove

File: pipeline/transform.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+# Datacleaning
+# Match mention to topic_id - will need to get topic_id from rds
+# Sentiment analysis
```
Comment (SurinaCS) [inline pipeline/transform.py:17]: maybe remove these or put in script docstring?

File: pipeline/transform.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+logging.basicConfig(
+    level=logging.INFO,
+    format='%(asctime)s - %(levelname)s - %(message)s',
+    handlers=[
+        logging.StreamHandler(),
+    ]
+)
+
+
+def get_cursor(config):
```
Comment (RidwanHamid501) [inline pipeline/transform.py:28]: typehints

File: pipeline/transform.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    logging.info("Loading raw data from test_content_data.csv")
+    content_dataframe = pd.read_csv(
+        "bluesky_output_data/bluesky_output_20241203_110911.csv")
+    keywords = extract_keywords_from_csv(
+        "bluesky_output_data/bluesky_output_20241203_110911.csv")
+
+    cleaned_dataframe = clean_data(content_dataframe, keywords)
+    keyword_map = ensure_keywords_in_db(keywords, cursor, conn)
+    matched_dataframe = keyword_matching(cleaned_dataframe, keyword_map)
+    final_dataframe = add_sentiment_scores(matched_dataframe)
```
Comment (SurinaCS) [inline pipeline/transform.py:120]: can this be moved into a main function?

File: pipeline/transform.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    """Find and add the sentiment scores of each message."""
+    analyzer = SentimentIntensityAnalyzer()
+
+    bluesky_data['sentiment_score'] = bluesky_data['text'].apply(
+        lambda text: analyzer.polarity_scores(text)['compound'])
+
+    return bluesky_data
+
+
+def extract_keywords_from_csv(csv_file):
```
Comment (SurinaCS) [inline pipeline/transform.py:99]: typehints please!

**Review by RidwanHamid501 — CHANGES_REQUESTED** [review]
please include typehints

File: pipeline/transform.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+logging.basicConfig(
+    level=logging.INFO,
+    format='%(asctime)s - %(levelname)s - %(message)s',
+    handlers=[
+        logging.StreamHandler(),
+    ]
+)
+
+
+def get_connection(config: dotenv_values) -> conn:
```
Comment (SurinaCS) [inline pipeline/transform.py:22]: great type hints

**Review by SurinaCS — APPROVED** [review]
nice changes made

**Pull request merged**